### PR TITLE
Add singular_values, svd_U, and svd_V

### PR DESCRIPTION
**Pull request by Dashadower**
## Summary

This implements `singular_values`, `svd_U`, and `svd_V`, with its respective gradients based on Eigen's [JacobiSVD](https://eigen.tuxfamily.org/dox/classEigen_1_1JacobiSVD.html)

## Tests
```
test/unit/math/rev/fun/singularvalues_test.cpp
test/unit/math/prim/fun/svd_U_test.cpp
test/unit/math/prim/fun/svd_V_test.cpp
test/unit/math/mix/fun/svd_U_test.cpp
test/unit/math/mix/fun/svd_V_test.cpp
```

## Side Effects

`svd_U` and `svd_V` may have reversed column-wise signs from other SVD library's results.

## Release notes

Implement `svd_U` and `svd_V`, add gradients to `singular_values`.

## Checklist

- [x] Math issue #2192 

- [x] Copyright holder: Hyunji Moon, Ben Bales, Shinyoung Kim

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/rev/fun/singular_values.hpp
Comment (bbbales2) [inline stan/math/rev/fun/singular_values.hpp:33]: I think you only need to get the thinU and thinV here.

File: stan/math/prim/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/prim/fun/svd_U.hpp:21]: @SteveBronder @andrjohns I'm struggling on this one -- at this point most of the matrix code accepts matrices of size zero and doesn't throw? Is that what we want in the future? We could definitely make this function return 0x0 matrices and not error.

Or I guess in general when should the matrix code throw and when should it just accept size zero?matrices?

File: stan/math/rev/fun/singular_values.hpp
Comment (bbbales2) [inline stan/math/rev/fun/singular_values.hpp:38]: [optional] Change these variables to `arena_U` and `arena_V` so it's easy to see where they live.

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:35]: Integers don't need copied to the arena since they do not allocate memory on the heap.

File: stan/math/rev/fun/singular_values.hpp
Comment (bbbales2) [inline stan/math/rev/fun/singular_values.hpp:29]: Don't need to save this

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:8]: Not used

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:11]: Not used

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:12]: Add this into `stan/math/prim/fun/Eigen.hpp`

File: stan/math/prim/fun/singular_values.hpp
Comment (bbbales2) [inline stan/math/prim/fun/singular_values.hpp:21]: It looks like this returns matrices of size zero instead of throwing. We'll need to make this consistent across all the functions but let's wait for a response to the question in `prim/fun/svd_U.hpp` first.

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:20]: I don't think this is necessary information.

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:22]: Change to:

```
Adjoint update equation comes from Equation (4)...
```

File: stan/math/rev/fun/svd_U.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_U.hpp:43]: Do `arena_t<Eigen:MatrixXd> arena_Fp(M, M);`

And then just assign stuff directly there and you don't have to copy it later.

File: stan/math/rev/fun/svd_V.hpp
Comment (bbbales2) [inline stan/math/rev/fun/svd_V.hpp:1]: Same comment as `svd_U`.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
The singular values prim tests need more in there. Just use the same matrices you test w/ svd_U and svd_V. We apparently were lazy when we implemented this (https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/fun/singular_values_test.cpp).

Also add a 2x3 matrix to the mix singular_values test.

There's another set of changes to make this support the new varmat matrix type (https://github.com/stan-dev/design-docs/blob/master/designs/0005-static-matrices.md).

I think all the comments I left here are pretty easy to straighten out (we'll wait for some feedback on the matrices). Ping me after you go through them, and I'll add instructions on doing the varmat stuff. It's another bunch of small changes. I've apparently left 14 comments here which is enuff stuff for one review tho :D.

**Comment by Dashadower** [comment]
@bbbales2 Changed as you requested. There already seems to be a 2x3 matrix test in `test/unit/math/mix/fun/singular_values_test.cpp`.

File: stan/math/prim/fun/singular_values.hpp
Comment (bbbales2) [reply stan/math/prim/fun/singular_values.hpp:21]: Just make the prim version throw I think. That'll be simplest for now. We can roll it back. Later.

So instead of returning a 0x0, do: `check_nonzero_size("singular_values", "m", m);`

Comment (bbbales2) [inline stan/math/prim/fun/singular_values.hpp:21]: Alright so we'll do the varmat conversion here first. Along with `Eigen::Matrix<var, R, C>`, we just added a new autodiff type to Stan (this [one](https://github.com/stan-dev/design-docs/blob/master/designs/0005-static-matrices.md)). We don't have developer docs up so lemme walk you through how to convert this.

First change `require_not_vt_var<EigMat>` to `require_not_st_var<EigMat>`.

The first says "Require the value type of the template parameter to not be a var". Varmats are `var_value<Eigen::Matrix<T, R, C>>` types, and the value type of a `var_value<Eigen::Matrix<T, R, C>>` is `Eigen::Matrix<T, R, C>`. The scalar type though is `var`.

So for this to work with varmat we want to say "Require the scalar type of the template parameter to not be a var"

File: stan/math/rev/fun/singular_values.hpp
Comment (bbbales2) [inline stan/math/rev/fun/singular_values.hpp:23]: Change: `require_eigen_matrix_dynamic_t<EigMat>` to `require_rev_matrix_t<EigMat>`

That means accept `Eigen::Matrix<var, R, C>` and `var_value<Eigen::Matrix<double, R, C>>` types, not just the first.

Delete `require_vt_var<EigMat>`. Not necessary.

File: stan/math/rev/fun/singular_values.hpp
Comment (bbbales2) [inline stan/math/rev/fun/singular_values.hpp:26]:
```
using ret_type = return_var_matrix_t<Eigen::VectorXd, EigMat>;
```

This computes the autodiff return type for a non-autodiff type `Eigen::VectorXd` given the input argument is `EigMat`. If this is an `Eigen::Matrix<var, R, C>` type, the output will similarly be an Eigen matrix of vars (call this a matvar for short). If the input is a `var_value<Eigen::Matrix<double, R, C>>`, the output will be a `var_value<Eigen::VectorXd>` (call this a varmat for short). Both are valid vector autodiff types in Stan.

Comment (bbbales2) [reply stan/math/rev/fun/singular_values.hpp:26]: I think with that `singular_values` should work with varmats. Now we gotta add tests.

File: test/unit/math/mix/fun/svd_U_test.cpp
Comment (bbbales2) [inline test/unit/math/mix/fun/svd_U_test.cpp:10]: To do varmat tests, for every `stan::test::expect_ad(f, ...)` do a `stan::test::expect_ad_matvar(f, ...)`.

And that's that! This should work for `singular_values`, `svd_U`, and `svd_V`. Lemme know if you have problems.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Alright, the varmat conversion. I left directions in the comments. Lemme know what I forgot (cause I probably did and the error messages will inevitably be confusing)

**Comment by bbbales2** [comment]
@Dashadower I made a few edits, check em' [here](https://github.com/stan-dev/math/pull/2286/commits/02c5a81342d46579dff511c741bdd8a7165d11b0). If you're good with those, I'll merge this when the tests pass.

**Review by bbbales2 — APPROVED** [review]

**Pull request merged**
